### ISSUE TITLE
function argument is lowered incorrectly

### ISSUE THREAD

**Issue by schweitzpgi** [flang-compiler/f18-llvm-project#959]
```fortran
function f(f2, arg)
  real f, arg
  interface
    function f2(rarg)
      real f2, rarg
    end function
  end interface
  f = f2(arg)
end function
```
is lowered as
```mlir
func @_QPf(%arg0: () -> (), %arg1: !fir.ref<f32>) -> f32 {
  %0 = fir.alloca f32 {bindc_name = "f", uniq_name = "_QFfEf"}
  %1 = fir.call @_QPf2(%arg1) : (!fir.ref<f32>) -> f32
  fir.store %1 to %0 : !fir.ref<f32>
  %2 = fir.load %0 : !fir.ref<f32>
  return %2 : f32
}
func private @_QPf2(!fir.ref<f32>) -> f32
```

This is clearly incorrect as `f2` is a reference/pointer to a function and not the name of an actual function.

The lowering should be
```mlir
func @_QPf(%arg0: (!fir.ref<f32>) -> f32, %arg1: !fir.ref<f32>) -> f32 {
  %0 = fir.alloca f32 {bindc_name = "f", uniq_name = "_QFfEf"}
  %1 = fir.call %arg0(%arg1) : (!fir.ref<f32>) -> f32
  fir.store %1 to %0 : !fir.ref<f32>
  %2 = fir.load %0 : !fir.ref<f32>
  return %2 : f32
}
```

**Comment by jeanPerier** [comment]
There is a known semantics bug with dummy procedure that are also specified as interface. This leads to lowering handling them as if they were external functions. Fix upcoming in https://reviews.llvm.org/D106697.

**Comment by jeanPerier** [comment]
Was fixed by Cherry-picking #964.

Now lowers to:
```
func @_QPf(%arg0: () -> (), %arg1: !fir.ref<f32>) -> f32 {
  %0 = fir.alloca f32 {bindc_name = "f", uniq_name = "_QFfEf"}
  %1 = fir.convert %arg0 : (() -> ()) -> ((!fir.ref<f32>) -> f32)
  %2 = fir.call %1(%arg1) : (!fir.ref<f32>) -> f32
  fir.store %2 to %0 : !fir.ref<f32>
  %3 = fir.load %0 : !fir.ref<f32>
  return %3 : f32
}
```
